### PR TITLE
Stats - Remove unnecessary api calls from Insights

### LINKED ISSUE #17912: [iOS] Stats: Limit endpoint access for Insights view
Whenever the Insights view in Stats currently refreshes, we request data from every available Insights endpoint. To reduce network access and improve performance, we should try to only request data for Insights cards that the user has added to their view.

### PR DESCRIPTION
**Pull request by Gio2018**
Fixes #17912 

This PR reduces the api calls in Stats - Insights, limiting them to only those calls that are needed for the currently visible cards.

If a card is added (by either tapping on "Add Card" or on the cog in the upper right, and then choosing a card to add: see screenshot below), the cache is invalidated and a new call will be performed, fetching also the data for the newly added card.

<p align=center>

</p>

To test:
Pre-requisites: 
- use an account with one or more sites that have stats. 
- before building this branch, build and run trunk and take note of the stats of one or more sites, to compare them with what you will see on this build

Testing steps
- Build/run this branch
- Go to My Site -> Stats
- Select "Insights" from the tab bar controller (if not already there)
- Make sure the insights you're seeing are consistent with the data you'd expect
- Verify that only the calls necessary for the visible cards are performed. To do so, you can either:
    - Use Charles Proxy to monitor the calls (note that this PR only affects calls for Insights, any other call performed at the time of accessing Stats is yet to be optimized)
    - Look at logs in the console, in the form of `Stats: Insights - successfully fetched dddd` where `dddd` is the data type associated to one or more card(s).
    - If you use the console, you can also compare the calls that were made before this change by simply turning off the `statsPerformanceImprovements` feature flag
- Select another tab, wait for the cache to expire and then go back to Insights, and make sure the data are refreshed properly and that the number of calls is still what you expect to see
- Add and remove cards at will, and make sure that the number of calls performed is still consistent.

Tip: to reduce wait times for the cache to expire, you can tweak the following parameter at line 112 in `StatsInsightStore`

`fileprivate static let cacheTTL: TimeInterval = 300 // 5 minutes`

Note:

most cards have a 1:1 correspondence with api calls, but some have non-obvious data associated to them. Here's a list of exceptions (hopefully I did not miss any):

- the "Grow Audience" card requires `allTimeStats`
- "Most Popular" and "This Year" both use `annualAndMostPopular`
- "Followers" makes two api calls to retrieve `wpcom` and `email` followers
- "Followers total" needs the same data as "Followers" plus `publicize` 

the type `insightsDataForSection` is used to represent this mapping..

## Regression Notes
1. Potential unintended areas of impact
Stats - Insights

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Use a feature flag for the changes, so they can be tested extensively

3. What automated tests I added (or what prevented me from doing so)
none
PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. - Added `SiteStatsInformationTests`
- [x] ~~I have considered adding accessibility improvements for my changes.~~ not needed
- [x] ~~I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.~~ not needed for this change specifically. We might want to add a line in the RN explaining that we made a variety of performance improvements when we ship this release.

### CODE REVIEW

**Review by frosty — COMMENTED** [review]
Code changes look good to me! I'm seeing one issue during testing, which can be recreated as follows:

* Add a small set of cards and pull to refresh to see the various endpoints listed in the console.
* Remove one or more of the cards, and pull to refresh again.
* You should still see the endpoints being hit for the cards you removed.
* If you switch tabs and come back then pull to refresh, then the list of endpoints should be correct.

I _think_ the issue is that we need to do a `writeInsightsToUserDefaults` in `removeInsight`, as we do when adding an insight or the view disappearing.

File: WordPress/Classes/Stores/StatsInsightsStore.swift [hunk 1]
```diff
@@ -248,50 +258,165 @@ private extension StatsInsightsStore {
...
+            fetchLastPostSummary()
+            return
+        }
+        updateFetchingStatusForVisibleCards(.loading)
+        fetchInsightsCards()
     }
 
-    func fetchOverview() {
+    func fetchInsightsCards() {
+
```
Comment (sla8c) [inline WordPress/Classes/Stores/StatsInsightsStore.swift]: Nit - can remove this empty line

**Review by sla8c — COMMENTED** [review]
LGTM!  One suggestion from me; I like how you moved the retrieval and saving of userDefaults state into `SiteStatsInformation` out of the `SiteStatsInsightsTableViewController`.  It might be nice to have a unit test for `SiteStatsInformation` and these 2 functions since these are headless operations?

**Comment by Gio2018** [comment]
> Code changes look good to me! I'm seeing one issue during testing, which can be recreated as follows:
> 
> * Add a small set of cards and pull to refresh to see the various endpoints listed in the console.
> * Remove one or more of the cards, and pull to refresh again.
> * You should still see the endpoints being hit for the cards you removed.
> * If you switch tabs and come back then pull to refresh, then the list of endpoints should be correct.
> 
> I _think_ the issue is that we need to do a `writeInsightsToUserDefaults` in `removeInsight`, as we do when adding an insight or the view disappearing.

Awesome, thanks for finding this @frosty !!

**Comment by Gio2018** [comment]
@frosty @sla8c ready for another pass

**Review by frosty — APPROVED** [review]
Changes look good to me!

**Review by sla8c — APPROVED** [review]
Didn't think I needed to approve but just in case you were waiting for me I'm approving.  Test looks good thanks!

**Comment by Gio2018** [comment]
Thank you both for re-reviewing @sla8c & @frosty !

**Comment by Gio2018** [comment]
FYI I am also removing the property `userDefaultsInsightTypesKey` from `StatsInsightsTableViewController` which is not in use anymore.

**Pull request merged**